module/VuFind/src/VuFind/Feature/DirLocationsTrait.php

```diff
@@ -0,0 +1,105 @@
+<?php
+
+/**
+ * VuFind Dir Locations Trait - Used to create path stacks based on DirLocations.ini
+ *
+ * PHP version 8
+ *
+ * Copyright (C) Hebis Verbundzentrale 2024
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 2,
+ * as published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+ *
+ * @category VuFind
+ * @package  Feature
+ * @author   Thomas Wagener <wagener@hebis.uni-frankfurt.de>
+ * @license  http://opensource.org/licenses/gpl-2.0.php GNU General Public License
+ * @link     https://vufind.org Main Page
+ */
+
+namespace VuFind\Feature;
```

Comment:
Is it possible it's worth putting this in the VuFind\Config\Feature namespace, since it has to do with reading the directory configuration?

I also wonder if it might make sense to design this a little more like the IniReaderTrait, where it uses a helper object (e.g. something like \VuFind\Config\DirLocationsStackReader) in a property and initializes it when unset, so there's less business logic in the trait and it's easier to override that logic if desired. Maybe it would even make sense to make that a service so that there's a singleton of the reader and the stack can be centrally stored to avoid repeating the calculations in multiple contexts... though I wouldn't insist on that if it causes too much disruption in terms of factories, etc. (It could also be a future enhancement if we set the stage for it by making the design flexible to begin with).

Reply:
The DirLocations stack resolving needs to be used in multiple locations where no direct access to the service locator is given (LanguageInitializerTrait,VuFind/Config/Locator,PathResolverTrait, etc.). So we would have to refactor the workflows in some way. I'm not sure if that is worth the effort. Or am I missing an easy solution here?

And I don't understand how moving the business logic into a helper object would make it easier to override it. Wouldn't you still have to create your own trait and helper object and apply it everywhere on your own anyway?

I don't reject your suggestion and also think it would be good if this part would be overridable but just don't know how to accomplish it.

Reply:
The LanguageInitializerTrait is only used in two places. My concerns were also the other places where the stack is resolved. However, the VuFind/Config/Locator is deprecated and replaced by the PathResolver and PathResolverTrait is only used in tests. So we probably don't need the logic there. I don't know what XSLT/Processor.php is used for. We could also ignore that.

What do you think about adding a "getLocalConfigDirStack" getter to the PathResolver and just using that instead of a new service?

Reply:
I implemented this now as suggested by adding a "getLocalConfigDirStack" method to the PathResolver.